File: Xenon/Xenon/HomingMissile.ts
import CBullet = require("Bullet");
import gsCVector = require("Vector");
import gsCControls = require("Controls");
import gsCTimer = require("Timer");
import enums = require("Enums");

class CHomingMissile extends CBullet {

    //-------------------------------------------------------------

    HOMING_MISSILE_TURNRATE = 10.0;
    HOMING_MISSILE_STARTUP = 0.250;
    HOMING_MISSILE_THRESHOLD = 8.0;

    //-------------------------------------------------------------
    m_angle: number;
    m_target: gsCVector;
    m_has_target: boolean;

    constructor() {
        super();
        this.m_target = new gsCVector(0.0, 0.0);
        this.m_has_target = false;
        this.m_name = "HomingMissile";
    }

    //-------------------------------------------------------------

    public getActorInfo() {
        this.m_actorInfo = this.m_scene.GetlistOfActors();
        return this.m_actorInfo.GetActorInfoListItem(enums.ActorInfoType.INFO_HOMING_MISSILE);
    }

    //-------------------------------------------------------------

    public activate(): boolean {

        if (!this.isActive()) {
            if (this.m_velocity.Y < 0.0) {
                this.m_angle = 0.0;
            }
            else {
                this.m_angle = 180.0;
            }
            this.m_timer.start();
        }
        return super.activate();
    }

    //-------------------------------------------------------------

    public update(controls: gsCControls, gameTime: gsCTimer): boolean {

        var direction: number = 0;
        this.m_position.plusEquals(this.m_velocity);

        if (this.m_has_target && this.m_timer.getTime() >= this.HOMING_MISSILE_STARTUP) {

            var relpos: gsCVector = this.m_target.minus(this.m_position);

            if (relpos.length > this.HOMING_MISSILE_THRESHOLD) {

                var a: number = relpos.direction();
                var da: number = a - this.m_angle;

                if (this.gsAbs(da) < this.HOMING_MISSILE_TURNRATE) {
                    this.m_angle = a;
                }
                else {
                    while (da < 0.0) {
                        da += 360.0;
                    }
                    while (da >= 360.0) {
                        da -= 360.0;
                    }
                    if (da < 180.0) {
                        this.m_angle += this.HOMING_MISSILE_TURNRATE;
                    }
                    else {
                        this.m_angle -= this.HOMING_MISSILE_TURNRATE;
                    }
                }

                while (this.m_angle < 0.0) {
                    this.m_angle += 360.0;
                }
                while (this.m_angle >= 360.0) {
                    this.m_angle -= 360.0;
                }

                var temp: gsCVector = new gsCVector(0, 0);
                //var temp2: gsCVector = temp.polar(this.getActorInfo().m_speed[this.m_grade], this.m_angle);
                var temp2: gsCVector = temp.polar(5, this.m_angle);
                this.setVelocity(temp2);
            }
            else
                this.m_has_target = false;
        }

        direction = ((this.m_angle + 22.5) / 45.0) & 7;

        this.m_sprite.setFrame(/*(int) m_grade + */ direction);		//TEMP

        return true;
    }

    //-------------------------------------------------------------

    public setTarget(target: gsCVector): void {
        this.m_target = target;
        this.m_has_target = true;
    }

    //-------------------------------------------------------------

    private gsAbs(v: number): number {
        return v >= 0.0 ? v : -v;
    }

    //-------------------------------------------------------------
}

export = CHomingMissile;

File: Xenon-Original_C++_Code/xenon/source/homingmissile.cpp
//-------------------------------------------------------------
//
// Class:	CHomingMissile
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CBullet
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CHomingMissile::CHomingMissile()
{
	m_target = gsCVector(0.f,0.f);
	m_has_target = false;
}

//-------------------------------------------------------------

CHomingMissile::~CHomingMissile()
{
}

//-------------------------------------------------------------

bool CHomingMissile::activate()
{
	if (!isActive()) {
		if (m_velocity.getY() < 0.f)
			m_angle = 0.f;
		else
			m_angle = 180.f;
		m_timer.start();
		}

	return CActor::activate();
}

//-------------------------------------------------------------

bool CHomingMissile::update(Controls *controls)
{
	int direction = 0;

	m_position += m_velocity;

	if (m_has_target && m_timer.getTime() >= HOMING_MISSILE_STARTUP) {
	
		gsCVector relpos = m_target - m_position;

		if (relpos.length() > HOMING_MISSILE_THRESHOLD) {

			float a = relpos.direction();

			float da = a - m_angle;

			if (gsAbs(da) < HOMING_MISSILE_TURNRATE)
				m_angle = a;
			else {
				while (da < 0.f)
					da += 360.f;
				while (da >= 360.f)
					da -= 360.f;

				if (da < 180.f)
					m_angle += HOMING_MISSILE_TURNRATE;
				else
					m_angle -= HOMING_MISSILE_TURNRATE;
				}

			while (m_angle < 0.f)
				m_angle += 360.f;
			while (m_angle >= 360.f)
				m_angle -= 360.f;
			
			setVelocity(gsCVector::polar(getActorInfo().m_speed[m_grade],m_angle));
			}
		else
			m_has_target = false;
		}

	direction = ((int) ((m_angle + 22.5f) / 45.f)) & 7;

	m_sprite.setFrame(/*(int) m_grade + */ direction);		//TEMP

	return true;
}

//-------------------------------------------------------------

void CHomingMissile::setTarget(const gsCVector& target)
{
	m_target = target;

	m_has_target = true;
}

//-------------------------------------------------------------
